ResourceManagerImpl + docs

ajkannan opened the pull request:
Also fixes #360

mziccard commented:
(just being annoying here, feel free to skip this paragraph) In general it would be cool to avoid unnecessary merge commits: c992427 could have been avoided by rebasing your local `resource-manager-impl` on top of `resource-manager` before working on ca53a03 and sending the PR.

@aozarov If you have funnier stuff to do today I would be happy to give a first pass to this tomorrow morning.

aozarov commented:
Thanks @mziccard that helps!

ajkannan commented:
Rebased to remove merge commit

ajkannan commented:
@mziccard thanks for the pass!  I added a commit to include all your feedback thus far.

mziccard commented:
Only one more minor comment on my side. And an [open discussion](https://github.com/GoogleCloudPlatform/gcloud-java/pull/457#commitcomment-15024785) not necessarily related to this PR.

aozarov commented on TESTING.md:
```diff
@@ -65,5 +65,51 @@ Here is an example that clears the bucket created in Step 3 with a timeout of 5
+
+  This method will block until the server thread has been terminated.
+
+#### On a remote machine
```
current emulator only works for localhost, right?

ajkannan replied:
Yup, removing the "remote" section in next commit

aozarov commented on gcloud-java-resourcemanager/README.md:
````diff
@@ -27,27 +27,145 @@ Example Application
+import com.google.gcloud.resourcemanager.ProjectInfo;
+```
+
+Then add the following code to create a project (be sure to change `myProjectId` to be something unique).
````
to your own unique project id.

aozarov commented on gcloud-java-resourcemanager/README.md:
````diff
@@ -27,27 +27,145 @@ Example Application
+    .addLabel("launch-status", "in-development").build());
+```
+
+Note that the values of the project you pass in to `replace` overwrite the server's values for non-read-only fields, namely `projectName` and `labels`. For example, if you create a project with `projectName` "some-project-name" and subsequently call replace using a `ProjectInfo` object that didn't set the `projectName`, then the server will unset the project's name. The server ignores any attempt ...
````
BTW, does it retain the original createTime?

ajkannan replied:
Yup, it retains the original createTime

aozarov commented on gcloud-java-resourcemanager/README.md:
```diff
@@ -27,27 +27,145 @@ Example Application
+Note that the values of the project you pass in to `replace` overwrite the server's values for non-read-only fields, namely `projectName` and `labels`. For example, if you create a project with `projectName` "some-project-name" and subsequently call replace using a `ProjectInfo` object that didn't set the `projectName`, then the server will unset the project's name. The server ignores any attempt ...
+
+#### Listing all projects
+Suppose that we want a list of all projects for which we have read permissions. Add the following import:
```
Does it actually work for user with only read permissions? (considering that our auth is specifying admin scope).

ajkannan replied:
Yes, I can view read-only projects using "get" and "list" even though the oauth scope is read/write

aozarov commented on gcloud-java-resourcemanager/README.md:
```diff
@@ -27,27 +27,145 @@ Example Application
+
+#### Complete source code
+
+Here we put together all the code shown above into one program.  This program assumes that you are running from your own desktop.
```
Shall we mention something like "and used the gcloud sdk to authenticate yourself"?

aozarov commented on gcloud-java-resourcemanager/pom.xml:
```diff
@@ -27,6 +27,12 @@
       <artifactId>google-api-services-cloudresourcemanager</artifactId>
       <version>v1beta1-rev6-1.19.0</version>
       <scope>compile</scope>
+      <exclusions>
```
is that still an issue with the new api client?

ajkannan replied:
The cloudresourcemanager artifact uses version 1.19 of google-api-client, which pulls some version of guava-jdk5 (maybe 13, but I'd have to double check).  guava-jdk5 version 13 had an old version of clock, which caused build issues.  The latest api client verison is 1.21, which also pulls guava-jdk5, but it doesn't specify a specific version and I didn't check compatibility with clock.

aozarov commented on gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/ResourceManagerOptions.java:
```diff
@@ -29,14 +30,14 @@
   private static final long serialVersionUID = 538303101192527452L;
   private static final String GCRM_SCOPE = "https://www.googleapis.com/auth/cloud-platform";
   private static final Set<String> SCOPES = ImmutableSet.of(GCRM_SCOPE);
+  private static final String DEFAULT_HOST = "https://cloudresourcemanager.googleapis.com";
```
I am curious why this services uses a special host rather than the default "https://www.googleapis.com"?

ajkannan replied:
Unsure, I got that URL from the docs: https://cloud.google.com/resource-manager/reference/rest/

aozarov commented on gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/ResourceManagerImpl.java:
```diff
@@ -0,0 +1,235 @@
+    }
+  }
+
+  private abstract static class BasePageFetcher<T extends Serializable>
```
This is a common base class for different services.

Not for this PR (so probably creating an issue) but we should move it to core and use Generics for the specific Options type `O` and provide an abstract method to return the `PAGE_TOKEN` option.

aozarov commented on gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/ProjectInfo.java:
```diff
@@ -324,7 +323,7 @@ public Builder toBuilder() {
     }
     if (createTimeMillis != null) {
-      projectPb.setCreateTime(ISODateTimeFormat.dateTime().print(createTimeMillis));
+      projectPb.setCreateTime(ISODateTimeFormat.dateTime().withZoneUTC().print(createTimeMillis));
```
`ProjetInfoTest` didn't need to change?

ajkannan replied:
It was using our local time zone, so converting to and from milliseconds wasn't an issue.  I'll add a test that makes sure that the serialized time zone ends with a 'Z' for UTC

aozarov commented on gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/ProjectInfo.java:
```diff
@@ -31,7 +31,6 @@
 
 /**
  * A Google Cloud Resource Manager project metadata object.
- * 
```
If you want to make checkstyle happy add a <p> after a new line (no new line after the <p>).
Also, State's ACTIVE and LIFECYCLE_STATE_UNSPECIFIED are missing a '.' in their javadoc.

Maybe after this PR run google checkstyle on all module files.

aozarov commented:
Looks good. Few nits and I didn't see the changes we discussed verbally.

ajkannan commented:
Note: the latest commit addresses comments and also handles the issue of "Unnamed" being returned by "list" as the project name for projects with a null name.
